src/auto_diff.jl

```diff
@@ -38,11 +38,27 @@ function Yao.AD.backward_params!(st::Tuple{<:MajoranaReg,<:MajoranaReg},
     out, outδ = st
     ham = Yao.AD.generator(block)
     majoranaham = yaoham2majoranasquares(ham)
-    g = outδ.state ⋅ (majoranaham * out.state) / 4
-    pushfirst!(collector, g)
+    # multiply majoranaham with out.state and inner product with outδ.state
+    # i.e. g = outδ.state ⋅ (majoranaham * out.state) / 4
+    pushfirst!(collector, fast_overlap(outδ.state, majoranaham, out.state)/4)
     return nothing
 end
 
+# compute tr(y' * A * x)
+function fast_overlap(y::AbstractVecOrMat{T}, A::SparseMatrixCSC{T}, x::AbstractVecOrMat{T}) where T
```

Comment:
I think it could happen that `eltype(y) != eltype(A)`, e.g. if we want to do simulations with reduced precision but haven't specified the circuit with reduced precision. So maybe have the types here `y::AbstractVecOrmat{T1}, A::SparseMatrixCSC{T2}, x::AbstractVecOrMat{T2}) where {T1, T2}` and then `g = zero(promote_type(T1, T2))` later?